### ISSUE TITLE
tapir server to ignore any validations 

### ISSUE THREAD

**Issue by ShahOdin** [softwaremill/tapir#3682]
let's say my api is supposed to return two "happy" path scenarios:

- 200
- 204

I don't want to complicate my service logic with handling the 204 case. instead I want to model it as an error that my service logic throws so my service logic considers it an error whereas the customer happily sees it as 204. with the default setup, the opinionated tapir gets in the way and manipulates things and I end up giving me 400. 

Also, how can I throw my own error in case of the user not having provided a header as part of the security input for example? I don't want tapir to automatically craft a 401. 

Basically, how can I get tapir to be as dumb as possible and let me control everything dynamically with my (Error, Status)?

**Comment by adamw** [comment]
If I understand you correctly, using the `statusCode` [output](https://tapir.softwaremill.com/en/latest/endpoint/ios.html) should solve your problem. You can either use a fixed status code, e.g. `.errorOut(statusCode(StatusCode.Accepted))`, or a dynamic one, e.g. `.out(statusCode)`, then you'll have to provide the status code as part of the server logic's result.

**Comment by shah-ovo** [comment]
Thanks! I'll double check but think I was getting a 400 where I had explicitly asked for 204 using the status output as you mentioned. Will add a minimal example when I'm home.

What about the explicit error on the missing header that's currently handled by tapir?

**Comment by adamw** [comment]
I think a custom `DecodeFailureHandler` ([docs](https://tapir.softwaremill.com/en/latest/server/errors.html#decode-failures)) set via the server options should do the trick

**Comment by shah-ovo** [comment]
> I think a custom DecodeFailureHandler ([docs](https://tapir.softwaremill.com/en/latest/server/errors.html#decode-failures)) set via the server options should do the trick

Yeah this did it:

```scala
  def handleMissingAcceptHeader[F[_]: Applicative]: DecodeFailureHandler[F] =
    new DecodeFailureHandler[F]:
      override def apply(ctx: DecodeFailureContext)(implicit
          monad: MonadError[F]
      ): F[Option[ValuedEndpointOutput[_]]] = ctx.failingInput match
        case EndpointIO.Header("X-Api-Key", _, _) =>
          hardcoded401
        case _ => DefaultDecodeFailureHandler[F](ctx)
```

<details>
  <summary>with the hardcoded value being (click to expand): </summary>

```scala
val hardcoded401 = server.model
  .ValuedEndpointOutput(
    statusCode.and(headers).and(jsonBody[Unauthorised]),
    (
    StatusCode(401),
    List(
        Header("www-authenticate", "ApiKey"),
        Header("content-type", "text/plain; charset=UTF-8")
    ),
    Unauthorised(
        ErrorMessage("Authentication has failed or has not yet been provided.")
    )
    )
)
.some
.pure[F]
```  

</details>

thank you. 

As for the other issue, sorry it doesn't give you 400, but a 500. Here is a minimal example:

```scala
val serverOptions: Http4sServerOptions[IO] = Http4sServerOptions
      .customiseInterceptors[IO]
      .options

    endpoint
      .in("hello")
      .errorOut(statusCode)
      .errorOut(stringBody)
      .serverLogic(_ =>
        IO.pure(
          Left(
            (StatusCode(204), "I'm being silenced!")
          )
        )
      )
      .pipe(Http4sServerInterpreter[IO](serverOptions).toRoutes)
      .run(
        Request[IO](
          uri = Uri.unsafeFromString(
            "hello"
          )
        )
      )
      .value
      .unsafeRunSync()
      .get
      .status
```

gives you 500, and not 204. How can I override this?

**Comment by adamw** [comment]
Ah, if you take a look at the exception that is being thrown by the interpreter, it's because you are trying to respond with a 204 (no content), and set a body. If you don't return a body, then this works fine.

**Comment by shah-ovo** [comment]
Sure, but as a dev I want to override that! `http4s` lets me do that I think. Does `tapir` let me get around that?

**Comment by adamw** [comment]
I don't think so, though it would be a matter of removing this case: https://github.com/softwaremill/tapir/blob/e00bad225b5e8ec6d332ad9c720f18dada982e3a/server/core/src/main/scala/sttp/tapir/server/interpreter/ServerInterpreter.scala#L248

Is there a use-case where you want to send a body with a 204?

**Comment by ShahOdin** [comment]
> Is there a use-case where you want to send a body with a 204?

From the open api spec I have been given:

```json
"no_content": {
        "content": {
          "application/json": {
            "example": "No data exists for the requested time period",
            "schema": {
              "type": "string"
            }
          }
        }
```

**Comment by adamw** [comment]
Interesting :) Well, if that's an obstacle, can you create a PR which removes this check?

**Comment by gaeljw** [comment]
I'm late to the party but I think returning a body with a 204 is "forbidden" per the HTTP specification. The fact that some implementations let you do so is not compliant with the spec.

Whether Tapir should let you do it if the underlying implementation accept it, I don't know...

> A 204 response is terminated by the end of the header section; it cannot contain content or trailers.
> - https://www.rfc-editor.org/rfc/rfc9110#status.204

On the other hand, the specification is not as "imposing it" as much as it does for 205:
> Since the 205 status code implies that no additional content will be provided, a server MUST NOT generate content in a 205 response.

My 2 cents :)

**Comment by adamw** [comment]
@gaeljw Yeah I understand the arguments, but as it turns out this proves to be an obstacle in some cases. The "living standard" seems to have diverged from the RFC :). I guess we could still optionally perform the check using an interceptor, if needed.
